modules/git/repo_tag.go

```diff
@@ -161,37 +162,83 @@ func (repo *Repository) GetTag(name string) (*Tag, error) {
 
 // GetTagInfos returns all tag infos of the repository.
 func (repo *Repository) GetTagInfos(page, pageSize int) ([]*Tag, int, error) {
-	// TODO this a slow implementation, makes one git command per tag
-	stdout, err := NewCommand(repo.Ctx, "tag").RunInDir(repo.Path)
+	forEachRefFmt := foreachref.NewFormat("objecttype", "refname:short", "object", "objectname", "creator", "contents", "contents:signature")
+
+	stdout, err := NewCommand(repo.Ctx, "for-each-ref", "--format", forEachRefFmt.Flag(), "--sort", "-*creatordate", "refs/tags").RunInDir(repo.Path)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	tagNames := strings.Split(strings.TrimRight(stdout, "\n"), "\n")
-	tagsTotal := len(tagNames)
-
-	if page != 0 {
-		tagNames = util.PaginateSlice(tagNames, page, pageSize).([]string)
-	}
-
-	tags := make([]*Tag, 0, len(tagNames))
-	for _, tagName := range tagNames {
-		tagName = strings.TrimSpace(tagName)
-		if len(tagName) == 0 {
-			continue
+	var tags []*Tag
+	parser := forEachRefFmt.Parser(strings.NewReader(stdout))
+	for {
+		ref := parser.Next()
+		if ref == nil {
+			break
 		}
 
-		tag, err := repo.GetTag(tagName)
+		tag, err := parseTagRef(ref)
 		if err != nil {
-			return nil, tagsTotal, err
+			return nil, 0, fmt.Errorf("GetTagInfos: parse tag: %w", err)
 		}
-		tag.Name = tagName
 		tags = append(tags, tag)
 	}
+	if err := parser.Err(); err != nil {
+		return nil, 0, fmt.Errorf("GetTagInfos: parse output: %w", err)
+	}
+
+	tagsTotal := len(tags)
+	if page != 0 {
+		tags = util.PaginateSlice(tags, page, pageSize).([]*Tag)
+	}
+	// TODO shouldn't be necessary since we're running with '--sort'?
```

Comment:
note: a bit unsure about need to sort here, guidance appreciated.

Reply:
Looks like there is a bug here before. `Pagination` is before `sortTagsByTime`.

Reply:
I was considering skipping the `sortTagsByTime` call since I believe this is already handled by `--sort -*creatordate` which is passed to the command, but I'll add it as a safety measure. be07fd7f3